provisioning/ironic_deployment.go

```diff
@@ -0,0 +1,189 @@
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	appsclientv1 "k8s.io/client-go/kubernetes/typed/apps/v1"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ironicv1alpha1 "github.com/metal3-io/ironic-standalone-operator/api/v1alpha1"
+	metal3iov1alpha1 "github.com/openshift/cluster-baremetal-operator/api/v1alpha1"
+)
+
+const (
+	ironicDeploymentName     = "metal3-ironic"
+	ironicAPICredentialsName = ironicSecretName // Reuse existing secret name
+	ironicTrustedCAName      = externalTrustBundleConfigMapName
+)
+
+// parseDHCPRange parses a DHCP range string in the format "IP1,IP2" and returns the begin and end IPs.
+func parseDHCPRange(dhcpRange string) (begin, end string, err error) {
+	if dhcpRange == "" {
+		return "", "", nil
+	}
+
+	parts := strings.Split(dhcpRange, ",")
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("invalid DHCP range format: %q, expected format: \"IP1,IP2\"", dhcpRange)
+	}
+
+	return strings.TrimSpace(parts[0]), strings.TrimSpace(parts[1]), nil
+}
+
+// buildIronicSpec constructs an Ironic CR spec from the Provisioning CR.
+func buildIronicSpec(info *ProvisioningInfo) (ironicv1alpha1.IronicSpec, error) {
+	spec := ironicv1alpha1.IronicSpec{
+		APICredentialsName: ironicAPICredentialsName,
+		TLS: ironicv1alpha1.TLS{
+			CertificateName:        tlsSecretName,
+			DisableVirtualMediaTLS: info.ProvConfig.Spec.DisableVirtualMediaTLS,
+		},
+		DeployRamdisk: ironicv1alpha1.DeployRamdisk{
+			SSHKey: info.SSHKey,
+		},
+	}
+
+	// Configure networking
+	networking := ironicv1alpha1.Networking{}
+
+	if UseIronicProxy(info) {
+		// When the proxy is using the normal Ironic public port we need to force Ironic to use a private port.
+		networking.APIPort = ironicPrivatePort
+	}
+
+	// Set IP address based on provisioning network configuration
+	// When provisioning network is managed, use the provisioning IP
+	if info.ProvConfig.Spec.ProvisioningNetwork == metal3iov1alpha1.ProvisioningNetworkManaged {
```

Comment:
Using IPAddress is not limited to managed networking, I suggest always setting it if ProvisioningIP is not empty.